internal/docstore/docstore.go

```diff
@@ -0,0 +1,216 @@
+	driver driver.Collection
+}
+
+// NewCollection is intended for use by provider implementations.
+var NewCollection = newCollection
+
+// newCollection makes a Collection.
+func newCollection(d driver.Collection) *Collection {
+	return &Collection{driver: d}
+}
+
+// Actions returns an ActionList that can be used to perform
+// actions on the collection's documents.
+func (c *Collection) Actions() *ActionList {
+	return &ActionList{coll: c}
+}
+
+// An ActionList is a sequence of actions that affect a single collection.
+type ActionList struct {
+	coll    *Collection
+	actions []*Action
+}
+
+// An Action is a read or write on a single document.
+// Use the methods of ActionList to create and execute Actions.
+type Action struct {
+	kind       driver.ActionKind
+	doc        Document
+	fieldpaths []string // paths to retrieve, for Get
+	mods       Mods     // modifications to make, for Update
+}
+
+func (l *ActionList) add(a *Action) *ActionList {
+	l.actions = append(l.actions, a)
+	return l
+}
+
+// Create adds an action that creates a new document.
+// The document must not already exist; an AlreadyExists error is returned if it
+// does. If the document doesn't have key fields, it will be given key fields with
+// unique values.
+// TODO(jba): treat zero values for struct fields as not present?
+func (l *ActionList) Create(doc Document) *ActionList {
+	return l.add(&Action{kind: driver.Create, doc: doc})
+}
+
+// Replace adds an action that replaces a document.
+// The document must already exist; a NotFound error is returned if it does not.
+func (l *ActionList) Replace(doc Document) *ActionList {
+	return l.add(&Action{kind: driver.Replace, doc: doc})
+}
+
+// Put adds an action that adds or replaces a document.
+// The document may or may not already exist.
+func (l *ActionList) Put(doc Document) *ActionList {
+	return l.add(&Action{kind: driver.Put, doc: doc})
+}
+
+// Delete adds an action that deletes a document.
+// Only the key fields and RevisionField of doc are used.
+// If the document doesn't exist, nothing happens and no error is returned.
+func (l *ActionList) Delete(doc Document) *ActionList {
+	// Rationale for not returning an error if the document does not exist:
+	// Returning an error might be informative and could be ignored, but if the
+	// semantics of an action list are to stop at first error, then we might abort a
+	// list of Deletes just because one of the docs was not present, and that seems
+	// wrong, or at least something you'd want to turn off.
+	return l.add(&Action{kind: driver.Delete, doc: doc})
+}
+
+// Get adds an action that retrieves a document.
+// Only the key fields of doc are used to create the request.
+// If fps is omitted, all the fields of doc are set to those of the
+// retrieved document. If fps is present, only the given field paths are
+// set. In both cases, other fields of doc are not touched.
+//
+// A field path is a dot-separated sequence of UTF-8 field names. A field path
```

Comment:
Consider making a type for `FieldPath`, I suspect you use it again. Examples would help in the docstring.

Reply:
Done.